src/mongodb/table.rs

```diff
@@ -0,0 +1,234 @@
+use crate::mongodb::connection_pool::MongoDBConnectionPool;
+use crate::mongodb::utils::expression::{combine_exprs_with_and, expr_to_mongo_filter};
+use crate::mongodb::Error;
+use async_trait::async_trait;
+use datafusion::arrow::datatypes::SchemaRef;
+use datafusion::catalog::{Session, TableProvider};
+use datafusion::common::project_schema;
+use datafusion::error::{DataFusionError, Result as DataFusionResult};
+use datafusion::execution::TaskContext;
+use datafusion::logical_expr::{Expr, TableProviderFilterPushDown, TableType};
+use datafusion::physical_expr::EquivalenceProperties;
+use datafusion::physical_plan::execution_plan::{Boundedness, EmissionType};
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning, PlanProperties,
+    SendableRecordBatchStream,
+};
+use datafusion::sql::TableReference;
+use futures::TryStreamExt;
+use mongodb::bson::Document;
+use serde_json;
+use std::{any::Any, fmt, sync::Arc};
+
+#[derive(Debug)]
+pub struct MongoDBTable {
+    pool: Arc<MongoDBConnectionPool>,
+    schema: SchemaRef,
+    table_reference: Arc<TableReference>,
+}
+
+impl MongoDBTable {
+    pub async fn new(
+        pool: &Arc<MongoDBConnectionPool>,
+        table_reference: impl Into<TableReference>,
+    ) -> Result<Self, Error> {
+        let table_reference = table_reference.into();
+        let schema = pool.connect().await?.get_schema(&table_reference).await?;
+
+        Ok(Self {
+            pool: Arc::clone(pool),
+            schema,
+            table_reference: Arc::new(table_reference),
+        })
+    }
+}
+
+#[async_trait]
+impl TableProvider for MongoDBTable {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        _state: &dyn Session,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(MongoDBExec::new(
+            Arc::clone(&self.table_reference),
+            Arc::clone(&self.pool),
+            Arc::clone(&self.schema),
+            projection,
+            filters,
+            limit,
+        )?))
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> DataFusionResult<Vec<TableProviderFilterPushDown>> {
+        Ok(vec![TableProviderFilterPushDown::Exact; filters.len()])
```

Comment:
Instead of claiming we support all filters, we should only claim to support filters that can be parsed by `expr_to_mongo_filter` - that will still allow DataFusion to push as much down as it can, but then it can handle any other filters that mongo can't - instead of erroring like the current implementation would do.